Add content for FedCM extensions (#28585)
mdn/content: files/en-us/web/api/credentialscontainer/get/index.md

- The [Credential Management API](/en-US/docs/Web/API/Credential_Management_API) uses `get()` to authenticate using basic federated credentials or username/password credentials.
- The [Web Authentication API](/en-US/docs/Web/API/Web_Authentication_API) uses `get()` to authenticate or provide additional factors during MFA with public key credentials (based on asymmetric cryptography).

> - The [Federated Credential Management API (FedCM)](/en-US/docs/Web/API/FedCM_API) uses `get()` to authenticate with federated identity providers (IdPs).

Author: You are correc.t I've fixed this instance, and a couple of others I found.

## Federated Credential Management API

> The [Federated Credential Management API (FedCM)](/en-US/docs/Web/API/FedCM_API) provides a standard mechanism for identity providers to make identity federation services available on the web in a privacy-preserving way without relying on third-party cookies and redirects, and a JavaScript API for sites to sign in with those services. Check out the linked API landing page for more usage information.

Reviewer: ```suggestion
The [Federated Credential Management API (FedCM)](/en-US/docs/Web/API/FedCM_API) provides a standard mechanism for identity providers to make identity federation services available on the web in a privacy-preserving way without relying on third-party cookies and redirects, and a JavaScript API for sites to use federated login. Check out the linked API landing page for more usage information.
```
Perhaps avoid "sign in" as the context parameter allows other kinds of usage, like sign up.

Author: I totally see what you are getting at, although I do think that as a synonym of "sign in", "login" will possibly have the same issue. I've had a look through the docs, and used language like this (in this instance and in other places) to make it clear that the API can handle more use cases than just sign-in:

> ...and a JavaScript API enabling use of federated sign-in for purposes such as signing in or signing up to a website.

Does that sound OK?

Reviewer: Ah good point. How about federated authentication?

> The [Federated Credential Management API (FedCM)](/en-US/docs/Web/API/FedCM_API) provides a standard mechanism for identity providers to make identity federation services available on the web in a privacy-preserving way without relying on third-party cookies and redirects, and a JavaScript API enabling use of federated authentication for purposes such as signing in or signing up to a website. Check out the linked API landing page for more usage information.

Author: I did this one, but in a slightly different way:

"...provides a standard mechanism for identity providers (IdPs) to enable identity federation services in a privacy-preserving way without relying on third-party cookies and redirects. This includes a JavaScript API enabling the use of federated authentication for purposes such as signing in or signing up to a website."

The thing is, they are not two separate functions. They are really the same function.

> The [Federated Credential Management API (FedCM)](/en-US/docs/Web/API/FedCM_API) provides a standard mechanism for identity providers to make identity federation services available on the web in a privacy-preserving way without relying on third-party cookies and redirects, and a JavaScript API enabling use of federated authentication for purposes such as signing in or signing up to a website. Check out the linked API landing page for more usage information.

Reviewer 2: Got it

> The [Federated Credential Management (FedCM) API](/en-US/docs/Web/API/FedCM_API) provides a standard mechanism for identity providers to make identity federation services available on the web in a privacy-preserving way without relying on third-party cookies and redirects, and a JavaScript API enabling use of federated authentication for purposes such as signing in or signing up to a website. Check out the linked API landing page for more usage information.

Reviewer 2: ```suggestion
The [Federated Credential Management (FedCM) API](/en-US/docs/Web/API/FedCM_API) provides a standard mechanism for identity providers (IdPs) to enable identity federation services in a privacy-preserving way without relying on third-party cookies and redirects. This includes a JavaScript API that enables the use of federated authentication for purposes such as signing in or signing up to a website. For more usage information, check out the linked landing page for the API.
```

> **Note:** Usage of `get()` with the `identity` parameter may be blocked by an {{httpheader("Permissions-Policy/identity-credentials-get", "identity-credentials-get")}} [Permissions Policy](/en-US/docs/Web/HTTP/Permissions_Policy) set on your server.

### `identity` object structure

`identity` can contain the following properties:

- `context` {{optional_inline}}

>   - : An enumerated value specifying the context in which the user is authenticating with FedCM. The browser uses this to vary the text in its FedCM sign-in UI so that it better suits the context. Possible values are:

Author: Agreed. I've changed this (and other instances) in my next commit.

- `providers`
  - : An array of objects specifying details of the different IdPs to be used to sign in. Each object can contain the following properties:
    - `configURL`
      - : A string specifying the URL of the IdP's config file. See [Provide a config file](/en-US/docs/Web/API/FedCM_API#provide_a_config_file) for more information.
    - `clientId`
      - : A string specifying the RP's client identifier, issued by the IdP to the RP in a completely separate process specific to the IdP.
    - `loginHint` {{optional_inline}}
      - : A string that provides a hint as to which account option(s) the browser should provide for the user to sign in with. This is useful in cases where the user has already signed in and the site asks them to reauthenticate, which can be confusing when they have multiple accounts and can't remember which one they previously used. The value to use in the `loginHint` property can be taken from the user's previous sign-in, and is matched against the `login_hints` values provided by the IdP in the array of user information returned from the [accounts list endpoint](/en-US/docs/Web/API/FedCM_API#the_accounts_list_endpoint).

>     - `nonce` {{optional_inline}}

Author: I looked at the IDL; the ones that are not marked "required" I've assumed to be optional and marked as such in the docs. I think that's a reasonable inference?

The only weirdness in the spec is that `IdentityCredentialRequestOptions.providers` is not marked as required. It surely must be, as some of `IdentityProviderConfig`'s children are required, plus it won't really do much if you don't provide any `IdentityProviderConfig`. I therefore decided to not mark it as optional.

Let me know if you think that all sounds OK.

>     - `nonce` {{optional_inline}}

Reviewer: IDK why I thought the first two were not required... weird. But yes `providers` should be required and I have a PR up for that https://github.com/fedidcg/FedCM/pull/480. Let me ping it...

>     - `nonce` {{optional_inline}}

Author: And Sam merged it; perfect.

>       - : A random string that can be included to ensure the response is issued for this specific request, and prevent {{glossary("replay attack", "replay attacks")}}.

Author: It works on the final published page; see https://developer.mozilla.org/en-US/docs/Web/API/CredentialsContainer/get.

It is broken on the preview page because only the changed pages have previews generated. Any relative link to a page that wasn't changed will therefore 404.

Check out [Federated Credential Management API (FedCM)](/en-US/docs/Web/API/FedCM_API) for more details on how this works. This call will start off the sign-in flow described in [FedCM sign-in flow](/en-US/docs/Web/API/FedCM_API#fedcm_sign-in_flow).

A similar call including the `context` and `loginHint` extensions would look like so:

```js
async function signIn() {
  const identityCredential = await navigator.credentials.get({
    identity: {
      context: "signup",
      providers: [
        {
          configURL: "https://accounts.idp.example/config.json",
          clientId: "********",
          nonce: "******",
          loginHint: "user1@example.com",
        },
      ],
    },
  });
}
```

> > **Note:** Once a user has already signed in with an IdP, the IdP can call the static {{domxref("IdentityProvider.getUserInfo_static", "IdentityProvider.getUserInfo()")}} method on their return to get information about them. `getUserInfo()` has to be called from within an IdP-origin {{htmlelement("iframe")}} so that RP scripts cannot access the data. This information can then be used to provide a personalized welcome message and sign-in button. This pattern is already common on sites that use identity federation for sign-in, but `getUserInfo()` provides a way to achieve it without relying on third-party cookies.

Author: Suggestion committed, but unresolving so I can also think about the pronoun issue.

> > **Note:** Once a user has already signed in with an IdP, the IdP can call the static {{domxref("IdentityProvider.getUserInfo_static", "IdentityProvider.getUserInfo()")}} method on their return to get information about them. `getUserInfo()` has to be called from within an IdP-origin {{htmlelement("iframe")}} so that RP scripts cannot access the data. This information can then be used to provide a personalized welcome message and sign-in button. This pattern is already common on sites that use identity federation for sign-in, but `getUserInfo()` provides a way to achieve it without relying on third-party cookies.

Author: This was actually fairly tricky to rewrite. I settled on

"Once a user has already signed in with an IdP, the IdP can call the static {{domxref("IdentityProvider.getUserInfo_static", "IdentityProvider.getUserInfo()")}} method on the user's return to retrieve their details."

Is that Ok?

> > **Note:** Once a user has already signed in with an IdP, the IdP can call the static {{domxref("IdentityProvider.getUserInfo_static", "IdentityProvider.getUserInfo()")}} method on their return to get information about them. `getUserInfo()` has to be called from within an IdP-origin {{htmlelement("iframe")}} so that RP scripts cannot access the data. This information can then be used to provide a personalized welcome message and sign-in button. This pattern is already common on sites that use identity federation for sign-in, but `getUserInfo()` provides a way to achieve it without relying on third-party cookies.

Reviewer 2: Though it's a little change, it does makes the info clearer to me ('the user...their details'). Thanks for the update!

}
```

> > **Note:** Once a user has already signed in with an IdP, the IdP can call the static {{domxref("IdentityProvider.getUserInfo_static", "IdentityProvider.getUserInfo()")}} method on their return to get information about them. `getUserInfo()` has to be called from within an {{htmlelement("iframe")}} embedding a page from the IdP origin so that RP scripts cannot access the data. This information can then be used to provide a personalized welcome message and sign-in button. This pattern is already common on sites that use identity federation for sign-in, but `getUserInfo()` provides a way to achieve it without relying on third-party cookies.

Reviewer: The iframe wording is a bit weird. I think it can be 'an embedded idp origin <iframe>' or something like this? Right now it seems to ask for an iframe which itself embeds an idp iframe. What we want to say is just an iframe with the idp origin.

Author: Good point. In both cases, I've updated the wording to something along the lines of "This method has to be called from within an IdP-origin {{htmlelement("iframe")}} so that RP scripts cannot access the data."
